src/Sign.Core/DataFormatSigners/AggregatingSigner.cs

```diff
@@ -62,6 +62,38 @@ public async Task SignAsync(IEnumerable<FileInfo> files, SignOptions options)
             ArgumentNullException.ThrowIfNull(files, nameof(files));
             ArgumentNullException.ThrowIfNull(options, nameof(options));
 
+            if (options.RecurseContainers)
+            {
+                await SignContainersAsync(files, options);
+            }
+
+            // split by code sign service and fallback to default
+
+            var grouped = (from signer in _signers
+                           from file in files
+                           where signer.CanSign(file)
+                           group file by signer into groups
+                           select groups).ToList();
+
+            // get all files and exclude existing; 
+
+            // This is to catch PE files that don't have the correct extension set
+            var defaultFiles = files.Except(grouped.SelectMany(g => g))
+                                    .Where(_fileMetadataService.IsPortableExecutable)
+                                    .Select(f => new { _defaultSigner.Signer, f })
+                                    .GroupBy(a => a.Signer, k => k.f)
+                                    .SingleOrDefault(); // one group here
+
+            if (defaultFiles != null)
+            {
+                grouped.Add(defaultFiles);
+            }
+
+            await Task.WhenAll(grouped.Select(g => g.Key.SignAsync(g.ToList(), options)));
+        }
+
+        private async Task SignContainersAsync(IEnumerable<FileInfo> files, SignOptions options)
```

Comment:
Would `SignContainerContentsAsync` be more accurate?  Even if `RecurseContainers` is `false`, Sign CLI should still sign containers by default.